bpf-sys/src/type_gen.rs

```diff
@@ -0,0 +1,240 @@
+        let allowlist: &mut Vec<String> = if let Some(allowlist) = &mut self.allowlist {
+            allowlist
+        } else {
+            self.allowlist = Some(vec![]);
+            self.allowlist.as_mut().unwrap()
+        };
+        allowlist.push(pattern.to_string());
+        self
+    }
+
+    /// Dump BTF types as C source code into `outfile` including all the
+    /// necessary dependent types.
+    pub fn generate(self, outfile: impl AsRef<Path>) -> Result<()> {
+        let mut fobj = File::create(&outfile).or_else(|e| Err(TypeGenError::IO(e)))?;
+        let header_name = outfile.as_ref().file_name().unwrap();
+        let guard_name = header_name
+            .to_str()
+            .unwrap()
+            .to_uppercase()
+            .chars()
+            .map(|c| match c {
+                'A'..='Z' => c,
+                _ => '_',
+            })
+            .collect::<String>();
+        let guardstr = format!(
+            "#ifndef __{guard_name}__\n#define __{guard_name}__\n\n",
+            guard_name = guard_name
+        );
+        fobj.write_all(guardstr.as_bytes())
+            .or_else(|e| Err(TypeGenError::IO(e)))?;
+        fobj.flush().or_else(|e| Err(TypeGenError::IO(e)))?;
+        let mut rawfd = RawFdWrapper(fobj.into_raw_fd());
+        let white_re = if let Some(ref allowlist) = self.allowlist {
+            Some(RegexSet::new(allowlist).or(Err(TypeGenError::RegexError))?)
+        } else {
+            None
+        };
+        unsafe {
+            let dump_opts = {
+                let mut uninit = MaybeUninit::<btf_dump_opts>::zeroed();
+                (*uninit.as_mut_ptr()).ctx = &mut rawfd as *mut _ as *mut _;
+                uninit.assume_init()
+            };
+            let dumpptr = btf_dump__new(
+                self.btfptr,
+                ptr::null(),
+                &dump_opts as *const _,
+                Some(vdprintf_wrapper),
+            );
+            if (dumpptr as isize) < 0 {
+                return Err(TypeGenError::DumpError);
+            }
+            let dumpptr = BtfDumpWrapper(dumpptr);
+            for type_id in 1..=btf__get_nr_types(self.btfptr) {
+                let btftypeptr = btf__type_by_id(self.btfptr, type_id);
+                let nameptr = btf__name_by_offset(self.btfptr, (*btftypeptr).name_off);
+                if nameptr.is_null() {
+                    continue;
+                }
+
+                let cname = CStr::from_ptr(nameptr);
+                let namestr = cname.to_str().or(Err(TypeGenError::DumpError))?;
+                if let Some(wre) = &white_re {
+                    if !wre.is_match(namestr) {
+                        continue;
+                    }
+                }
+                if btf_dump__dump_type(dumpptr.0, type_id) < 0 {
+                    return Err(TypeGenError::DumpError);
+                }
+            }
+        }
+
+        let mut fobj = unsafe { File::from_raw_fd(rawfd.0) };
```

Comment:
Maybe we could do an `impl Into<File> for RawFdWrapper`, so there's RAII continuity between the two types the other way around. `let fobj: File = rawfd.into()` would then consume `rawfd`.

Reply:
Cool. Using `Into` trait looks elegant. I will try it.